signer/xpi/recommendation.go

```diff
@@ -0,0 +1,149 @@
+package xpi // import "go.mozilla.org/autograph/signer/xpi"
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+type Recommendation struct {
+	// AddOnID is the ID of the extension this recommendation is
+	// for. Must match the ID in the extension’s manifest.json
+	AddOnID string `json:"addon_id"`
+
+	// States is a list of strings for each state of an addon that
+	// firefox understands
+	States []string `json:"states"`
+
+	// Validity is a pair of timestamps to expire a recommendation
+	// after an appropriate amount of time, since the
+	// recommendation is for a given version of the addon and it
+	// will need to be reissued for new versions.
+	Validity map[string]time.Time `json:"validity"`
+
+	// SchemaVersion is a uint to allow gradual upgrades of the
+	// recommendation file
+	SchemaVersion int `json:"schema_version"`
+}
+
+// Recommend returns a Recommendation for param addonID with param states
+func Recommend(addonID string, states []string, notBefore, notAfter time.Time) *Recommendation {
+	rec := &Recommendation{
+		AddOnID: addonID,
+		States:  states,
+		Validity: map[string]time.Time{
+			"not_before": notBefore,
+			"not_after":  notAfter,
+		},
+		SchemaVersion: 1,
+	}
+	return rec
+}
+
+func validateValidityTime(t time.Time) error {
+	if t.Nanosecond() != 0 {
+		return errors.Errorf("xpi: time must not include nanoseconds")
+	}
+	if t.Location() != time.UTC {
+		return errors.Errorf("xpi: time must be in UTC tz")
+	}
+	return nil
+}
+
+func (r *Recommendation) Validate(allowedRecommendationStates map[string]bool) error {
+	if len(r.States) < 1 {
+		return errors.Errorf("xpi: recommendation must include at least one state")
+	}
+	for _, state := range r.States {
+		if _, ok := allowedRecommendationStates[state]; !ok {
+			return errors.Errorf("xpi: recommendation included the invalid state %s", state)
+		}
+	}
+
+	if len(r.Validity) != 2 {
+		return errors.Errorf("xpi: recommendation must include validity with not_before and not_after fields")
+	}
+	if _, ok := r.Validity["not_before"]; !ok {
+		return errors.Errorf("xpi: recommendation validity with missing not_before field")
+	}
+	if _, ok := r.Validity["not_after"]; !ok {
+		return errors.Errorf("xpi: recommendation validity with missing not_after field")
+	}
+	if !r.Validity["not_before"].Before(r.Validity["not_after"]) {
+		return errors.Errorf("xpi: recommendation validity not_after must be after not_before field")
+	}
```

Comment:
I think it would also make sense to refuse recommendation that are valid for less than, say, 24 hours

Reply:
Why? Doesn't seem like it's useful to an attacker.